packages/kbn-esql-validation-autocomplete/scripts/generate_function_definitions.ts

```diff
@@ -0,0 +1,383 @@
+            values: {
+              value: arg.value,
+            },
+          }
+        ),
+        location: arg.location,
+      });
+    }
+  }
+  return messages;
+}`;
+
+const dateDiffSuggestions = [
+  'year',
+  'quarter',
+  'month',
+  'week',
+  'day',
+  'hour',
+  'minute',
+  'second',
+  'millisecond',
+  'microsecond',
+  'nanosecond',
+];
+
+const dateDiffOptions = [
+  'year',
+  'years',
+  'yy',
+  'yyyy',
+  'quarter',
+  'quarters',
+  'qq',
+  'q',
+  'month',
+  'months',
+  'mm',
+  'm',
+  'dayofyear',
+  'dy',
+  'y',
+  'day',
+  'days',
+  'dd',
+  'd',
+  'week',
+  'weeks',
+  'wk',
+  'ww',
+  'weekday',
+  'weekdays',
+  'dw',
+  'hour',
+  'hours',
+  'hh',
+  'minute',
+  'minutes',
+  'mi',
+  'n',
+  'second',
+  'seconds',
+  'ss',
+  's',
+  'millisecond',
+  'milliseconds',
+  'ms',
+  'microsecond',
+  'microseconds',
+  'mcs',
+  'nanosecond',
+  'nanoseconds',
+  'ns',
+];
+
+/**
+ * Enrichments for function definitions
+ *
+ * This is the place to put information that is not provided by Elasticsearch
+ * and, hence, won't be present in the JSON file.
+ */
+const functionEnrichments: Record<string, RecursivePartial<FunctionDefinition>> = {
+  log10: {
+    validate: validateLogFunctions,
+  },
+  log: {
+    validate: validateLogFunctions,
+  },
+  date_diff: {
+    signatures: [
+      {
+        params: [{ literalOptions: dateDiffOptions, literalSuggestions: dateDiffSuggestions }],
+      },
+    ],
+  },
+  date_extract: {
+    signatures: [
+      {
+        // override the first param as type chrono_literal
+        params: [{ type: 'chrono_literal' }],
+      },
+    ],
+  },
+  date_trunc: {
+    signatures: [
+      {
+        // override the first param to be of type time_literal
+        params: [{ type: 'time_literal' }],
+      },
+    ],
+  },
+  mv_sort: {
+    signatures: new Array(6).fill({
+      params: [{}, { literalOptions: ['asc', 'desc'] }],
+    }),
+  },
+  coalesce: {
+    signatures: supportedFieldTypes
+      .map<FunctionDefinition['signatures']>((type) => [
+        {
+          params: [
+            {
+              name: 'first',
+              type,
+              optional: false,
+            },
+          ],
+          returnType: type,
+          minParams: 1,
+        },
+        {
+          params: [
+            {
+              name: 'first',
+              type,
+              optional: false,
+            },
+            {
+              name: 'rest',
+              type: 'boolean',
+              optional: true,
+            },
+          ],
+          returnType: type,
+          minParams: 1,
+        },
+      ])
+      .flat(),
+  },
+  mv_dedupe: {
+    signatures: supportedFieldTypes.map<FunctionDefinition['signatures'][number]>((type) => ({
+      params: [
+        {
+          name: 'field',
+          type,
+          optional: false,
+        },
+      ],
+      returnType: type,
+    })),
+  },
```

Comment:
Fixes for both of these are planned in https://github.com/elastic/elasticsearch/issues/108982